File: library/src/main/java/com/getstream/sdk/chat/navigation/destinations/AttachmentDestination.java
package com.getstream.sdk.chat.navigation.destinations;

import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;

import com.getstream.sdk.chat.R;
import com.getstream.sdk.chat.model.ModelType;
import com.getstream.sdk.chat.utils.Utils;
import com.getstream.sdk.chat.utils.frescoimageviewer.ImageViewer;
import com.getstream.sdk.chat.view.activity.AttachmentActivity;
import com.getstream.sdk.chat.view.activity.AttachmentDocumentActivity;
import com.getstream.sdk.chat.view.activity.AttachmentMediaActivity;

import java.util.ArrayList;
import java.util.List;

import io.getstream.chat.android.client.logger.ChatLogger;
import io.getstream.chat.android.client.models.Attachment;
import io.getstream.chat.android.client.models.Message;

public class AttachmentDestination extends ChatDestination {

    public final Message message;
    public final Attachment attachment;
    public final String type;
    public final String url;

    public AttachmentDestination(Message message, Attachment attachment, Context context) {
        super(context);
        this.message = message;
        this.attachment = attachment;

        String url = "";
        String type = attachment.getType();

        switch (attachment.getType()) {
            case ModelType.attach_file:
            case ModelType.attach_video:
                url = attachment.getAssetUrl();
                break;
            case ModelType.attach_image:
                if (attachment.getOgUrl() != null) {
                    url = attachment.getOgUrl();
                    type = ModelType.attach_link;
                } else {
                    //multiple images case, no single url
                }
                break;
            case ModelType.attach_giphy:
                url = attachment.getThumbUrl();
                break;
            case ModelType.attach_product:
                url = attachment.getUrl();
                break;
        }

        this.url = url;
        this.type = type;
    }

    @Override
    public void navigate() {
        showAttachment(message, attachment);
    }

    public void showAttachment(Message message, Attachment attachment) {
        String url = null;
        String type = null;

        switch (attachment.getType()) {
            case ModelType.attach_file:
                loadFile(attachment);
                return;
            case ModelType.attach_image:
                if (attachment.getOgUrl() != null) {
                    url = attachment.getOgUrl();
                    type = ModelType.attach_link;
                } else {
                    List<String> imageUrls = new ArrayList<>();
                    for (Attachment a : message.getAttachments()) {
                        if (!a.getType().equals(ModelType.attach_image) || TextUtils.isEmpty(a.getImageUrl()))
                            continue;
                        imageUrls.add(a.getImageUrl());
                    }
                    if (imageUrls.isEmpty()) {
                        Utils.showMessage(context, "Invalid image(s)!");
                        return;
                    }

                    int position = message.getAttachments().indexOf(attachment);
                    if (position > imageUrls.size() - 1) position = 0;
                    new ImageViewer.Builder<>(context, imageUrls)
                            .setStartPosition(position)
                            .show();
                    return;
                }
                break;
            case ModelType.attach_video:
                url = attachment.getAssetUrl();
                break;
            case ModelType.attach_giphy:
                url = attachment.getAssetUrl();
                break;
            case ModelType.attach_product:
                url = attachment.getUrl();
                break;
        }
        if (TextUtils.isEmpty(url)) {
            Utils.showMessage(context, context.getString(R.string.stream_attachment_invalid_url));
            return;
        }

        if (type == null) type = attachment.getType();
        Intent intent = new Intent(context, AttachmentActivity.class);
        intent.putExtra("type", type);
        intent.putExtra("url", url);
        start(intent);
    }

    private void loadFile(Attachment attachment) {

        String mimeType = attachment.getMimeType();
        String url = attachment.getAssetUrl();

        if (mimeType == null) {
            ChatLogger.Companion.getInstance().logE("AttachmentDestination", "MimeType is null for url" + url);
            Utils.showMessage(context, context.getString(R.string.stream_attachment_invalid_mime_type, attachment.getName()));
        } else {
            // Media
            if (mimeType.contains("audio") ||
                    mimeType.contains("video")) {
                Intent intent = new Intent(context, AttachmentMediaActivity.class);
                intent.putExtra(AttachmentMediaActivity.TYPE_KEY, mimeType);
                intent.putExtra(AttachmentMediaActivity.URL_KEY, url);
                start(intent);
            } else if (mimeType.equals("application/msword") ||
                    mimeType.equals(ModelType.attach_mime_txt) ||
                    mimeType.equals(ModelType.attach_mime_pdf) ||
                    mimeType.contains("application/vnd")) {

                Intent intent = new Intent(context, AttachmentDocumentActivity.class);
                intent.putExtra("url", url);
                start(intent);
            }
        }
    }
}

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/common/navigation/destinations/AttachmentDestination.kt
package io.getstream.chat.android.ui.common.navigation.destinations

import android.content.Context
import android.content.Intent
import android.widget.Toast
import com.getstream.sdk.chat.model.ModelType
import com.getstream.sdk.chat.navigation.destinations.ChatDestination
import com.getstream.sdk.chat.utils.frescoimageviewer.ImageViewer
import com.getstream.sdk.chat.view.activity.AttachmentDocumentActivity
import io.getstream.chat.android.client.logger.ChatLogger
import io.getstream.chat.android.client.models.Attachment
import io.getstream.chat.android.client.models.Message
import io.getstream.chat.android.ui.common.R
import io.getstream.chat.android.ui.gallery.AttachmentActivity
import io.getstream.chat.android.ui.gallery.AttachmentMediaActivity

public open class AttachmentDestination(
    public var message: Message,
    public var attachment: Attachment,
    context: Context,
) : ChatDestination(context) {

    override fun navigate() {
        showAttachment(message, attachment)
    }

    public fun showAttachment(message: Message, attachment: Attachment) {
        if (attachment.type == ModelType.attach_file ||
            attachment.type == ModelType.attach_video ||
            attachment.mimeType?.contains(VIDEO_MIME_TYPE_PREFIX) == true
        ) {
            loadFile(attachment)
            return
        }

        var url: String? = null
        var type: String? = attachment.type

        when (attachment.type) {
            ModelType.attach_image -> {
                when {
                    attachment.ogUrl != null -> {
                        url = attachment.ogUrl
                        type = ModelType.attach_link
                    }
                    attachment.isGif() -> {
                        url = attachment.imageUrl
                        type = ModelType.attach_giphy
                    }
                    else -> {
                        showImageViewer(message, attachment)
                        return
                    }
                }
            }
            ModelType.attach_giphy -> url = attachment.thumbUrl
            ModelType.attach_product -> url = attachment.url
        }

        if (url.isNullOrEmpty()) {
            Toast.makeText(context, context.getString(R.string.stream_attachment_invalid_url), Toast.LENGTH_SHORT)
                .show()
            return
        }

        val intent = Intent(context, AttachmentActivity::class.java).apply {
            putExtra("type", type)
            putExtra("url", url)
        }
        start(intent)
    }

    private fun loadFile(attachment: Attachment) {
        val mimeType = attachment.mimeType
        val url = attachment.assetUrl

        if (mimeType == null) {
            ChatLogger.instance.logE("AttachmentDestination", "MimeType is null for url $url")
            Toast.makeText(
                context,
                context.getString(R.string.stream_attachment_invalid_mime_type, attachment.name),
                Toast.LENGTH_SHORT
            ).show()
            return
        }

        // Media
        when {
            mimeType.contains("audio") || mimeType.contains(VIDEO_MIME_TYPE_PREFIX) -> {
                val intent = Intent(context, AttachmentMediaActivity::class.java).apply {
                    putExtra(AttachmentMediaActivity.TYPE_KEY, mimeType)
                    putExtra(AttachmentMediaActivity.URL_KEY, url)
                }
                start(intent)
            }
            mimeType == ModelType.attach_mime_doc ||
                mimeType == ModelType.attach_mime_txt ||
                mimeType == ModelType.attach_mime_pdf ||
                mimeType == ModelType.attach_mime_html ||
                mimeType.contains("application/vnd") -> {
                val intent = Intent(context, AttachmentDocumentActivity::class.java).apply {
                    putExtra("url", url)
                }
                start(intent)
            }
        }
    }

    protected open fun showImageViewer(
        message: Message,
        attachment: Attachment,
    ) {
        val imageUrls: List<String> = message.attachments
            .filter { it.type == ModelType.attach_image && !it.imageUrl.isNullOrEmpty() }
            .mapNotNull(Attachment::imageUrl)

        if (imageUrls.isEmpty()) {
            Toast.makeText(context, "Invalid image(s)!", Toast.LENGTH_SHORT).show()
            return
        }

        val attachmentIndex = message.attachments.indexOf(attachment)

        ImageViewer.Builder(context, imageUrls)
            .setStartPosition(
                if (attachmentIndex in imageUrls.indices) attachmentIndex else 0
            )
            .show()
    }

    private fun Attachment.isGif() = mimeType?.contains("gif") ?: false

    private companion object {
        private const val VIDEO_MIME_TYPE_PREFIX = "video"
    }
}
